### PR TITLE
Improve ledger fetching

### PR DESCRIPTION
**Pull request by JoelKatz**
Fix a bug in SHAMapNodeID, remove some ledger fetching behavior that hurts, improve InboundLedgers, and improve how we fetch historical ledgers.

### CODE REVIEW

**Comment by nbougalis** [comment]
LGTM :+1:

**Comment by vinniefalco** [comment]
Note that the off-by-one commit message was rewritten in the proposed release

**Comment by nbougalis** [comment]
Noted. I will be force-pushing an update to account for this.

**Comment by nbougalis** [comment]
@vinniefalco: the message has been updated

**Comment by MarkusTeufelberger** [comment]
Building history seems to be broken between -b7 and -b8 by the way (rippled stops fetching older ledgers), does this fix it or is it a seperate issue?

**Comment by nbougalis** [comment]
More reviewers are needed.

**Comment by nbougalis** [comment]
@MarkusTeufelberger which version stops? `b7` or `b8`?

**Comment by MarkusTeufelberger** [comment]
b7 works, b8 stops (same config with ledger_history:full). I have downgraded for now, just a heads-up that there might be something going on.

**Comment by rec** [comment]
Looking at this now!

**Comment by nbougalis** [comment]
LGTM :+1:

**Comment by rec** [comment]
:+1:

File: src/ripple/app/ledger/LedgerMaster.cpp [hunk 1]
```diff
@@ -567,15 +568,23 @@ class LedgerMasterImp
 
     /** Request a fetch pack to get the ledger prior to 'nextLedger'
```
Comment (ximinez) [inline src/ripple/app/ledger/LedgerMaster.cpp]: This comment needs to be updated to match the updated function parameters.

**Comment by ximinez** [comment]
I don't yet understand all of the implications of all of the changes, but the changes look good, and I'm not seeing the race condition related error messages on this branch. :+1:

**Pull request closed without merging**